packages/astro/src/@types/astro.ts

```diff
@@ -421,11 +429,25 @@ export interface AstroUserConfig {
 // export interface AstroUserConfig extends z.input<typeof AstroConfigSchema> {
 // }
 
+type InjectedScriptStage = 'before-hydration' | 'head-inline' | 'page';
 /**
  * Resolved Astro Config
  * Config with user settings along with all defaults filled in.
  */
-export type AstroConfig = z.output<typeof AstroConfigSchema>;
+export interface AstroConfig extends z.output<typeof AstroConfigSchema> {
+	// Public:
+	// This is a more detailed type than zod validation gives us.
+	// TypeScript still confirms zod validation matches this type.
+	integrations: AstroIntegration[];
+	// Private:
+	// We have a need to pass context based on configured state,
+	// that is different from the user-exposed configuration.
+	// TODO: Create an AstroConfig class to manage this, long-term.
+	_ctx: {
```

Comment:
I don't know if I will die on this hill (I might) but I think mixing configuration and state is a mistake that will be hard to unravel. Probably the *right* thing to do is create a class/object that wraps the config and then pass that everywhere (I believe that is what ResolvedConfig is for).

*But* that would involve updating a crazy amount of code, so I understand not wanting to do that.

How about a compromise where this context object is kept in a global WeakMap?

```ts
interface Context { ... }

const contexts = new WeakMap<AstroConfig, Context>();
const getContext = (config: AstroConfig) => contexts.get(config);
```

Reply:
Yea, I tried to leave the TODO to capture that need here. I agree, a `context` object isn't really a property of your config. If anything, it's the opposite, so maybe `AstroContext` becomes the class that manages your config and other stuff. 

If we're talking whats in scope for this PR and the short-term, I think I prefer the `_ctx` hack vs. maintaining a weakmap separately.  It's not a hill that I'll die on either, but I don't see that as any easier/harder to unravel than `_ctx`.

The compromise that I'd suggest instead would be to fast-track that larger Class/Object refactor instead of letting this sit for to long post-merge. If we could commit to that (maybe @bholmesdev could take a look next week?) then I think `_ctx` is the better short-term path.